db/db.go

```diff
@@ -245,9 +208,34 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
 	return newData, newDataChecksum, output, isWrite, nil
 }
 
+// NewTransaction returns a new Transaction.
+func (db *DB) NewTransaction() *Transaction {
+	return &Transaction{
+		db:   db,
+		head: db.head,
```

Comment:
i don't think we're using db.head in a safe way:
1. we lock when we write to it but do not lock when we read it
2. we read db.head multiple times in a function and assume that it has not changed eg here in this example (db.head could change between setting head: and setting me:)

seems like we should have (not necessarily as part of this pr, just soon) a getter and a setter for head that both use the lock, functions should read the value exactly once, and all the code should stop fiddling with db.head directly. (this applies in diffserver as well)

cc @aboodman

Reply:
The second one is definitely true, good catch, scary. The first one doesn't seem dangerous to me but maybe i'm not seeing it.

Reply:
the first one is definitely true as well: setting a value is not guaranteed to be atomic so setting db.head while someone is reading db.head could has undefined behavior. i think we'd have to use something like https://golang.org/pkg/sync/atomic/#LoadPointer to guarantee a safe read if we are not locking.

Reply:
https://github.com/rocicorp/replicache/issues/42